File: day02/dive.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.34);
use experimental qw(signatures);

my @cmds;
while (<>) {
    chomp;
    my ($cmd, $val) = split / /;
    push @cmds, [$cmd, $val];
}

my ($d, $h, $a) = (0,0,0);
for (@cmds) {
    my ($cmd, $val) = $_->@*;
    if ($cmd eq 'forward') {
        $h += $val;
    } elsif ($cmd eq 'down') {
        $d += $val;
    } elsif ($cmd eq 'up') {
        $d -= $val;
    }
}
say "Part 1: ", $d * $h;

($d, $h, $a) = (0,0,0);
for (@cmds) {
    my ($cmd, $val) = $_->@*;
    if ($cmd eq 'forward') {
        $h += $val;
        $d += $a * $val;
    } elsif ($cmd eq 'down') {
        $a += $val;
    } elsif ($cmd eq 'up') {
        $a -= $val;
    }
}
say "Part 2: ", $d * $h;

File: day02/dive.py
import sys

cmds = []
with open(sys.argv[1]) as f:
    for line in f:
        cmd, val = line.rstrip().split(' ')
        cmds.append((cmd, int(val)))

d,h = 0,0
for cmd,val in cmds:
    match(cmd):
        case 'forward':
            h += val
        case 'down':
            d += val
        case 'up':
            d -= val
print('Part 1:', d * h)

d,h,a = 0,0,0
for cmd,val in cmds:
    match(cmd):
        case 'forward':
            h += val
            d += a * val
        case 'down':
            a += val
        case 'up':
            a -= val
print('Part 2:', d * h)
